workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowAndroidXSupport.kt

```diff
@@ -15,13 +27,162 @@ public object WorkflowAndroidXSupport {
    * Tries to get the parent lifecycle from the current view via [ViewTreeLifecycleOwner], if that
    * fails it looks up the context chain for a [LifecycleOwner], and if that fails it just returns
    * null.
+   *
+   * By explicitly looking at the context if the view tree is missing an owner, we can automatically
+   * handle cases where a custom Activity, for example, has overridden setContentView and thus
+   * disabled the ViewTreeOwner integration.
    */
   @WorkflowUiExperimentalApi
   public fun lifecycleOwnerFromViewTreeOrContext(view: View): LifecycleOwner? =
     ViewTreeLifecycleOwner.get(view) ?: view.context.lifecycleOwnerOrNull()
 
+  /**
+   * Tries to get the parent [SavedStateRegistryOwner] from the current view tree's view via
+   * [ViewTreeSavedStateRegistryOwner], if that fails it looks up the context chain for a one,
+   * and if that fails it just returns null.
+   *
+   * By explicitly looking at the context if the view tree is missing an owner, we can automatically
+   * handle cases where a custom Activity, for example, has overridden setContentView and thus
+   * disabled the ViewTreeOwner integration.
+   */
+  @WorkflowUiExperimentalApi
+  public fun savedStateRegistryOwnerFromViewTreeOrContext(view: View): SavedStateRegistryOwner? =
+    ViewTreeSavedStateRegistryOwner.get(view) ?: view.context.savedStateRegistryOwnerOrNull()
+
+  /**
+   * Computes a string key for using with a `SavedStateRegistry` that will be globally unique.
+   * Requires View IDs to be set for any views that are siblings and have the same rendering type.
+   * The key is generated by combining the rendering [compatibility keys][Named.keyFor] of each
+   * rendering starting with the current one and going up the view tree to the root. Each rendering
+   * key is also associated with the ID of the [View] it's tagged on.
+   */
+  @WorkflowUiExperimentalApi
+  public fun compositeViewIdKey(view: View): String {
+    return generateSequence(view) { it.parent as? ViewGroup }
+      // This will generate the list of ancestor IDs from child -> parent, so it will be in the
+      // reverse order we want (child/parent/root).
+      .map { it.id.toString() }
```

Comment:
Faster to do this after list conversion? I guess doesn't matter unless its parallelized for the list.

Reply:
if we did it after the list conversion, that would be two list traversals instead of one. Of course, it probably doesn't matter anyway since this list will unlikely ever be longer than 20-ish elements.